File: src/test/java/leetcode/LC035SearchInsertPosition.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/search-insert-position/
 */
public class LC035SearchInsertPosition {
    public class Solution {
        public int searchInsert(int[] nums, int target) {
            int a = 0, z = nums.length;
            while (a != z) {
                int m = (a + z) / 2;
                if (nums[m] < target) {
                    a = m + 1;
                    continue;
                }
                if (nums[m] >= target) {
                    z = m;
                    continue;
                }
            }
            return z;
        }
    }

    @Test
    public void test__empty__5() throws Exception {
        int[] nums = {};
        assertEquals(0, new Solution().searchInsert(nums, 5));
    }

    @Test
    public void test__1_3_5_6__5() throws Exception {
        int[] nums = {1, 3, 5, 6};
        assertEquals(2, new Solution().searchInsert(nums, 5));
    }

    @Test
    public void test__1_3_5_6__2() throws Exception {
        int[] nums = {1, 3, 5, 6};
        assertEquals(1, new Solution().searchInsert(nums, 2));
    }

    @Test
    public void test__1_3_5_6__7() throws Exception {
        int[] nums = {1, 3, 5, 6};
        assertEquals(4, new Solution().searchInsert(nums, 7));
    }

    @Test
    public void test__1_3_5_6__0() throws Exception {
        int[] nums = {1, 3, 5, 6};
        assertEquals(0, new Solution().searchInsert(nums, 0));
    }
}

File: algorithms/code/leetcode/lc035_search_insert_position/lc035_search_insert_position.rs
// https://leetcode.com/problems/search-insert-position/
// #easy

pub fn search_insert(nums: Vec<i32>, target: i32) -> usize {
    let mut a = 0usize;
    let mut z = nums.len();
    while a != z {
        let m = (a + z) >> 1;
        if nums[m] < target {
            a = m + 1;
        } else {
            z = m;
        }
    }
    z
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_and_5() {
        let nums: Vec<i32> = vec![];
        assert_eq!(0usize, search_insert(nums, 5));
    }

    #[test]
    fn test_1356_and_5() {
        let nums: Vec<i32> = vec![1, 3, 5, 6];
        assert_eq!(2usize, search_insert(nums, 5));
    }

    #[test]
    fn test_1356_and_2() {
        let nums: Vec<i32> = vec![1, 3, 5, 6];
        assert_eq!(1usize, search_insert(nums, 2));
    }

    #[test]
    fn test_1356_and_7() {
        let nums: Vec<i32> = vec![1, 3, 5, 6];
        assert_eq!(4usize, search_insert(nums, 7));
    }

    #[test]
    fn test_1356_and_0() {
        let nums: Vec<i32> = vec![1, 3, 5, 6];
        assert_eq!(0usize, search_insert(nums, 0));
    }

    #[test]
    fn test_1356_and_1() {
        let nums: Vec<i32> = vec![1, 3, 5, 6];
        assert_eq!(0usize, search_insert(nums, 1));
    }
}
